cfgrammar/src/lib/span.rs

```diff
@@ -43,3 +43,56 @@ impl Span {
         self.len() == 0
     }
 }
+
+pub struct NewlineToLineColCache<'a> {
```

Comment:
Hmm, here's something interesting. When I first (naively) thought about this, I assumed that the cache wouldn't need a copy of the source. That's wrong, of course, because `offset_to_line_col` needs that to calculate column numbers.

But now I'm wondering if/how this could support incrementally built-up newline caches (e.g. for a streaming parser). Any thoughts?

Reply:
I had considered it but (not in the streaming context though, i'm just not sure what type the streaming context would/should give for src exactly),

The way I as considering was moving the building of `newlines` into `offset_to_line_col` and checking the `off` parameter against `newlines.last()` in an a branch (after the `self.src.len()` check but before the line/col calculation. then using `extend` on `newlines`, using `self.src[last..off].char_indices.map(|cidx| cidx.o + last)`.

or something to that effect, the danger here is that off needs to be on an `str::is_char_boundary` for that slice to work I believe.  I don't see anywhere that the current mechanism relies on that, but it seems like it might be a reasonable requirement.

Reply:
Well, i've been thinking this morning about this and noticed that there is one issue with the sort of lazy building of the newlines vector, specifically I think in cases where a file ends without a newline, it'll try and lazily find more newlines for any offset on the last line.

This would seem like it could be an issue for code like minified javascript and json in such cases where you have excessively long lines.

In the non-streaming case we could just add an index into newlines at EOF (which besides variable name should be harmless)... That kind of seems like a poor approximation for just keeping a specific `usize` around for the furthest byte we've looked for a newline at.